# Размещение баннеров

Представим, что международное круизное агентство Carnival Cruise Line решило себя разрекламировать с помощью баннеров и обратилось для этого к вам. Чтобы протестировать, велика ли от таких баннеров польза, их будет размещено всего 20 штук по всему миру. Вам надо выбрать 20 таких локаций для размещения, чтобы польза была большой и агентство продолжило с вами сотрудничать.

Агентство крупное, и у него есть несколько офисов по всему миру. Вблизи этих офисов оно и хочет разместить баннеры — легче договариваться и проверять результат. Также эти места должны быть популярны среди туристов.

In [16]:
import numpy as np
import pandas as pd
from sklearn.cluster import MeanShift

Для поиска оптимальных мест воспользуемся базой данных крупнейшей социальной сети, основанной на локациях — Foursquare.

Часть открытых данных есть, например, на сайте archive.org:

https://archive.org/details/201309_foursquare_dataset_umn

Скачайте любым удобным образом архив fsq.zip с этой страницы.

Нас будет интересовать файл checkins.dat.

Для удобной работы с этим документом преобразуем его к формату csv, удалив строки, не содержащие координат — они неинформативны для нас:

In [17]:
#Reading the data file
checkins = pd.read_csv('checkins.dat', sep='|', skipinitialspace=True, skiprows=[1], low_memory=False)
print(checkins.info())
checkins.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1021967 entries, 0 to 1021966
Data columns (total 6 columns):
 #   Column            Non-Null Count    Dtype  
---  ------            --------------    -----  
 0   id                1021967 non-null  object 
 1   user_id           1021966 non-null  float64
 2   venue_id          1021966 non-null  float64
 3   latitude          396634 non-null   float64
 4   longitude         396634 non-null   float64
 5   created_at        1021966 non-null  object 
dtypes: float64(4), object(2)
memory usage: 46.8+ MB
None


id       user_id   venue_id   latitude        longitude       \
0  984301   2041916.0     5222.0             NaN             NaN   
1  984222     15824.0     5222.0       38.895112      -77.036366   
2  984315   1764391.0     5222.0             NaN             NaN   
3  984234     44652.0     5222.0       33.800745      -84.410520   
4  984249   2146840.0     5222.0             NaN             NaN   

      created_at        
0  2012-04-21 17:39:01  
1  2012-04-21 17:43:47  
2  2012-04-21 17:37:18  
3  2012-04-21 17:43:43  
4  2012-04-21 17:42:58

In [18]:
#Stripping columns names (deleting whitespaces)
checkins.columns = checkins.columns.str.strip()

#Dropping rows with NaN values
checkins = checkins.dropna()
print('Shape without NaN: %s' % str(checkins.shape))
checkins = checkins.iloc[:10**5, 3:-1]
checkins.head()

Shape without NaN: (396634, 6)


latitude   longitude
1   38.895112  -77.036366
3   33.800745  -84.410520
7   45.523452 -122.676207
9   40.764462 -111.904565
10  33.448377 -112.074037

Теперь необходимо кластеризовать данные координаты, чтобы выявить центры скоплений туристов. Поскольку баннеры имеют сравнительно небольшую площадь действия, нам нужен алгоритм, позволяющий ограничить размер кластера и не зависящий от количества кластеров.

Эта задача — хороший повод познакомиться с алгоритмом MeanShift. Его описание при желании можно посмотреть в sklearn user guide. Используйте MeanShift, указав bandwidth=0.1, что в переводе из градусов в метры колеблется примерно от 5 до 10 км в средних широтах.

Примечание: на 396634 строках кластеризация будет работать долго. Быть очень терпеливым не возбраняется — результат от этого только улучшится. Но для того, чтобы сдать задание, понадобится сабсет из первых 100 тысяч строк. Это компромисс между качеством и затраченным временем. Обучение алгоритма на всём датасете занимает около часа, а на 100 тыс. строк — примерно 2 минуты, однако этого достаточно для получения корректных результатов.

Некоторые из получившихся кластеров содержат слишком мало точек — такие кластеры не интересны рекламодателям. Поэтому надо определить, какие из кластеров содержат, скажем, больше 15 элементов. Центры этих кластеров и являются оптимальными для размещения.

Как мы помним, 20 баннеров надо разместить близ офисов компании. Найдем на Google Maps по запросу Carnival Cruise Line адреса всех офисов:

33.751277, -118.188740 (Los Angeles)

25.867736, -80.324116 (Miami)

51.503016, -0.075479 (London)

52.378894, 4.885084 (Amsterdam)

39.366487, 117.036146 (Beijing)

-33.868457, 151.205134 (Sydney)

Осталось определить 20 ближайших к ним центров кластеров. Т.е. посчитать дистанцию до ближайшего офиса для каждой точки и выбрать 20 с наименьшим значением.

In [19]:

def distance(x, y):
    return np.sqrt( np.sum((x - y)**2) )


In [20]:

def get_valid_clusters(data: pd.DataFrame) -> list:
    clustering = MeanShift(bandwidth=0.1).fit(checkins)
    unique_elements, count = np.unique(clustering.labels_, return_counts=True)
    valid_clusters = np.vstack((unique_elements, count))
    valid_clusters = valid_clusters[:, valid_clusters[1, :] > 15]
    valid_clusters = valid_clusters[0, :]
    return [cluster for ind, cluster in enumerate(clustering.cluster_centers_) if ind in valid_clusters]

offices_centers = [
    [33.751277, -118.188740],
    [25.867736, -80.324116],
    [51.503016, -0.075479],
    [52.378894, 4.885084],
    [39.366487, 117.036146],
    [-33.868457, 151.205134]
]
    
valid_clusters = get_valid_clusters(checkins)
with_min_valid_clusters = []
for valid_cluster in valid_clusters:
    min_val = np.inf
    for office in offices_centers:
        cur_val = distance(office, valid_cluster)
        if min_val > cur_val:
            min_val = cur_val
    with_min_valid_clusters.append([valid_cluster, min_val])
    
sorted_min_valid_clusters = sorted(with_min_valid_clusters, key=lambda x: x[1])
closest_point = sorted_min_valid_clusters[0]


[array([-33.86063043, 151.20477593]), np.float64(0.007834758163107856)]


Выберите из получившихся 20 центров тот, который наименее удален от ближайшего к нему офиса. Ответ в этом задании — широта и долгота этого центра, записанные через пробел.

In [22]:
coords, dist = closest_point
print("Coords: ", coords)
print("Dist: ", dist)


Coords:  [-33.86063043 151.20477593]
Dist:  0.007834758163107856
